stabilizer/src/main/java/com/hazelcast/stabilizer/provisioner/Provisioner.java

```diff
@@ -343,6 +345,40 @@ public void download(String dir) {
         echoImportant("Finished Downloading Artifacts of %s machines", addresses.size());
     }
 
+    public List<String> getPublicAdresses(List<AgentAddress> addresses){
+        List<String> result = new LinkedList<String>();
+        for(int i = 0; i < addresses.size(); i++){
+            result.add(addresses.get(i).publicAddress);
+        }
+        return result;
+    }
+
+    public void upload() throws IOException {
+        if(uploadPath != null){
```

Comment:
I coding tip, although it is a personal preference that helps to deal with checkstyle CyclomaticComplexity issues as well.

If you enter a method and there is an error/nothing to do, jump out of it ASAP.

So instead of 
```
if(somecondition){
}else{
}
```

Just do a:
```
if(somecondition){
    return 
    or
    throw BlaException();
}

.. rest of logic
```

This approach prevents you from doing a lot of deep nesting.